selfdrive/ui/replay/replay.cc

```diff
@@ -25,37 +72,280 @@ void Replay::parseResponse(QString response){
-    unlogger->process(sm);
+    process(sm);
   });
   thread->start();
 
-  QObject::connect(unlogger, &Unlogger::loadSegment, [=](){
-    addSegment(++current_segment);
+  seek_thread = new QThread;
+  QObject::connect(seek_thread, &QThread::started, [=](){
+    seekThread();
+  });
+  seek_thread->start();
+
+  queue_thread = new QThread;
+  QObject::connect(queue_thread, &QThread::started, [=](){
+    seekRequestThread();
   });
+  queue_thread->start();
+}
+
+void Replay::seekTime(int seek_){
+  // TODO: see if eidx also needs to be cleared
+  if(!seeking){
+    if(seek >= 0){
+      setSeekRequest(seek_*1e9);
+    }
+
+    if(seek_/60 != current_segment) {
+      int last_segment = current_segment;
+      current_segment = seek_/60;
+
+      for(int i = 0 ; i < 2*window_padding + 1 ; i++) {
+        // add segments that don't overlap
+        int seek_ind = seek_/60 - window_padding + i;
+        if(((last_segment + window_padding < seek_ind) || (last_segment - window_padding > seek_ind)) && (seek_ind >= 0)) {
+          addSegment(seek_ind);
+        }
+        // remove current segments that don't overlap
+        int cur_ind = last_segment - window_padding + i;
+        if(((seek_/60 + window_padding < cur_ind) || (seek_/60 - window_padding > cur_ind)) && (cur_ind >= 0)) {
+          trimSegment(cur_ind);
+        }
+      }
+    }
+  }
+}
+
+void Replay::seekRequestThread() {
+  while(1) {
+    if(seek_queue.size() > 0 && !seeking) {
+      int calculated_time = getRelativeCurrentTime()/1e9;
+      while(seek_queue.size() > 0) {
+        auto add_val = seek_queue.dequeue();
+        if(add_val.first) {
+          calculated_time += add_val.second;
+        } else {
+          calculated_time = add_val.second;
+        }
+      }
+      seekTime(calculated_time);
+    }
+  }
+}
+
+void Replay::seekThread(){
+  char c;
+  while(1){
+    c = getch();
+    if(c == '\n'){
+      printf("Enter seek request: ");
+      std::string request;
+      std::cin >> request;
+
+      seek_queue.clear();
+
+      if(request[0] == '#') {
+        request.erase(0, 1);
+        seek_queue.enqueue({false, std::stoi(request)*60});
+        continue;
+      }
+      seek_queue.enqueue({false, std::stoi(request)});
+      getch(); // remove \n from entering seek
+    } else if (c == 'm') {
+      seek_queue.enqueue({true, 60});
+    } else if (c == 'M') {
+      seek_queue.enqueue({true, -60});
+    } else if (c == 's') {
+      seek_queue.enqueue({true, 10});
+    } else if (c == 'S') {
+      seek_queue.enqueue({true, -10});
+    } else if (c == 'G') {
+      seek_queue.clear();
+      seek_queue.enqueue({false, 0});
+    } else if (c == ' ') {
+      togglePause();
+    }
+  }
+}
+
+void Replay::process(SubMaster *sm) {
+
+  active = true;
+
+  qDebug() << "hello from unlogger thread";
+  while (events.size() == 0) {
+    qDebug() << "waiting for events";
+    QThread::sleep(1);
+  }
+  qDebug() << "got events";
+
+  route_t0 = events.firstKey();
```

Comment:
get this from initData's logMonoTime. it's in all the segments

Reply:
This is what it's doing